### PR TITLE
Pull #40

### PR DESCRIPTION
**Pull request by Lioncat2002**
Issue #40 is complete
confirmTransaction moved to new issue to handle the changes required to implement it

### CODE REVIEW

**Comment by Lioncat2002** [comment]
Current need some clearance regarding the doctests and the default url

File: lib/solana.cpp [hunk 1]
```diff
@@ -67,6 +67,19 @@ json Connection::sendTransactionRequest(
   return jsonRequest("sendTransaction", params);
 }
 
+json Connection::setAirdrop(
```
Comment (mschneider) [inline lib/solana.cpp:70]: should be called requestAirdropRequest so it's not in line with the remaining code

File: lib/solana.cpp [hunk 2]
```diff
@@ -151,6 +163,71 @@ std::string Connection::signAndSendTransaction(
...
+}
+
+std::string Connection::requestAirdrop(const PublicKey &pubkey,uint64_t lamports,std::string url){
+
+  const json req = setAirdrop(pubkey.toBase58(),lamports);
+  cpr::Response r =
+      cpr::Post(cpr::Url{url}, cpr::Body{req.dump()},
+                cpr::Header{{"Content-Type", "application/json"}});
+  //json res = json::parse(r.text);
+  std::cout<<pubkey.toBase58()<<r.status_code << std::endl;
```
Comment (mschneider) [inline lib/solana.cpp:221]: library code should not just print on the log

File: lib/solana.cpp [hunk 3]
```diff
@@ -151,6 +163,71 @@ std::string Connection::signAndSendTransaction(
 
   return b58Sig;
 }
+
+std::string Connection::sendRawTransaction(
```
Comment (mschneider) [inline lib/solana.cpp:167]: duplicate code with signAndSendTransaction, please re-use methods

File: lib/solana.cpp [hunk 4]
```diff
@@ -151,6 +163,71 @@ std::string Connection::signAndSendTransaction(
...
+  
+  json res=r.text;
+  return res["result"];
+}
+
+std::string Connection::requestAirdrop(const PublicKey &pubkey,uint64_t lamports,std::string url){
+
+  const json req = setAirdrop(pubkey.toBase58(),lamports);
+  cpr::Response r =
+      cpr::Post(cpr::Url{url}, cpr::Body{req.dump()},
```
Comment (mschneider) [inline lib/solana.cpp:218]: connection has it's own url defined, no need to override it

File: tests/main.cpp [hunk 5]
```diff
@@ -3,6 +3,14 @@
 
 #include "MangoAccount.hpp"
 
+TEST_CASE("Request Airdrop"){
+    const auto fromKey = solana::PublicKey::fromBase58(
+      "83astBRguLMdt2h5U1Tpdq5tjFoJ6noeGwaY3mDLVcri");
+      auto connection = solana::rpc::Connection();
+  auto result = connection.requestAirdrop(fromKey,1000000000,"http://localhost:8899");
```
Comment (mschneider) [inline tests/main.cpp:10]: connection has it's own url defined, no need to override it when calling request Airdrop

File: lib/solana.cpp [hunk 6]
```diff
@@ -151,6 +163,71 @@ std::string Connection::signAndSendTransaction(
...
+                cpr::Header{{"Content-Type", "application/json"}});
+  //json res = json::parse(r.text);
+  std::cout<<pubkey.toBase58()<<r.status_code << std::endl;
+  if (r.status_code != 200)
+    throw std::runtime_error("unexpected status_code " +
+                             std::to_string(r.status_code));
+
+  
+  return r.text;
```
Comment (mschneider) [inline lib/solana.cpp:227]: response is probably json, make sure to parse it, same way other methods do it

**Comment by aniketfuryrocks** [comment]
Pending test cases :
- `sendTransaction`
Pending implementations :
- `simulateTransaction`
- `confirmTransaction`

**Comment by aniketfuryrocks** [comment]
Pending test cases
- `confirmTransaction`

Pending implementations
- `confirmTransaction`

File: tests/main.cpp [hunk 7]
```diff
@@ -1,8 +1,43 @@
...
+  const std::string memo = "Hello";
+
+  const solana::CompiledInstruction ix = {
+      1, {}, std::vector<uint8_t>(memo.begin(), memo.end())};
+
+  const solana::CompiledTransaction tx = {
+      recentBlockHash, {keyPair.publicKey, memoProgram}, {ix}, 1, 0, 1};
+
+  // call simulateTransaction
+  connection.simulateTransaction(keyPair, tx);
```
Comment (mschneider) [inline tests/main.cpp:29]: verify return value

File: tests/main.cpp [hunk 8]
```diff
@@ -1,8 +1,43 @@
...
+  // simulate transaction
+  connection.sendTransaction(keyPair, tx);
+}
+
+TEST_CASE("Request Airdrop") {
+  const solana::Keypair keyPair = solana::Keypair::fromFile(KEY_PAIR_FILE);
+  auto connection = solana::rpc::Connection(DEVNET);
+  auto result = connection.requestAirdrop(keyPair.publicKey, 1000000000);
+  CHECK_GT(result.length(), 0);
```
Comment (mschneider) [inline tests/main.cpp:38]: requestAirdrop should return the transaction signature of the aidrop request, not an empty response. also the wallet balance should increase

File: include/solana.hpp [hunk 9]
```diff
@@ -275,20 +404,60 @@ class Connection {
...
+  std::string sendTransaction(
+      const Keypair &keypair, const CompiledTransaction &tx,
+      const SendTransactionConfig &config = SendTransactionConfig()) const;
+
+  /**
+   * Send a transaction that has already been signed and serialized into the
+   * wire format
+   */
+  std::string sendRawTransaction(
+      const std::string &transaction,
```
Comment (mschneider) [inline include/solana.hpp:435]: std::vector<uint8_t>

File: include/solana.hpp [hunk 10]
```diff
@@ -275,20 +404,60 @@ class Connection {
...
+   * wire format, and encoded as a base64 string
+   */
+  std::string sendEncodedTransaction(
+      const std::string &transaction,
+      const SendTransactionConfig &config = SendTransactionConfig()) const;
+
+  /**
+   * Simulate sending a transaction
+   */
+  json simulateTransaction(const Keypair &keypair,
```
Comment (mschneider) [inline include/solana.hpp:449]: please return all data the json rpc api provides: [SimulatedTransactionResponse](https://solana-labs.github.io/solana-web3.js/modules.html#SimulatedTransactionResponse)

File: include/solana.hpp [hunk 11]
```diff
@@ -230,6 +232,33 @@ struct CompiledTransaction {
       instruction.serializeTo(buffer);
     }
   };
+
+  /**
+   * signs and base64 encodes the transaction
+   */
+  static std::string signAndEncodeTransaction(const Keypair &keypair,
```
[the remark this answers is not in the document]
Comment (Lioncat2002) [reply include/solana.hpp:239]: wouldn't it be better if it returns a json?

File: include/solana.hpp (hunk 9 above)
Comment (Lioncat2002) [reply include/solana.hpp]: fixed

File: include/solana.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (mschneider) [reply include/solana.hpp:239]: no just a binary blob vector u8 - signed and encoded transaction are usually just udp packets

[the remark this answers is not in the document]
Comment (Lioncat2002) [reply include/solana.hpp:239]: Ahh I see

**Comment by mschneider** [comment]
things still todo before merge:

- implement getBalance check
- update README / examples

File: tests/main.cpp (hunk 8 above)
Comment (Lioncat2002) [reply tests/main.cpp]: Fixed.
P.S. had to add a 15 sec wait in the test so that the validators can process the airdrop

File: tests/main.cpp [hunk 12]
```diff
@@ -58,9 +59,13 @@ TEST_CASE("Simulate & Send Transaction") {
 TEST_CASE("Request Airdrop") {
   const solana::Keypair keyPair = solana::Keypair::fromFile(KEY_PAIR_FILE);
   auto connection = solana::rpc::Connection(DEVNET);
+
+  auto prev_sol = connection.getBalance(keyPair.publicKey);
   auto result = connection.requestAirdrop(keyPair.publicKey, 1000000000);
-  // TODO: validate using confirmTransaction 
-  CHECK_GT(result.length(), 0);
+  std::this_thread::sleep_for(std::chrono::seconds(15));
```
Comment (mschneider) [inline tests/main.cpp:65]: please use confirmTransaction for this

Comment (mschneider) [reply tests/main.cpp:65]: oh i see in the next issue, that works with me

File: tests/main.cpp [hunk 13]
```diff
@@ -1,8 +1,73 @@
...
+#include <string>
+#include <thread>
+
+#include "solana.hpp"
 #define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
 #include <doctest/doctest.h>
 
 #include "MangoAccount.hpp"
 
+const std::string DEVNET = "https://api.devnet.solana.com";
```
Comment (mschneider) [inline tests/main.cpp:12]: move this definition to solana.hpp

https://github.com/mschneider/solcpp/pull/42/files/6934148919f55fb4e612bb519aa7dc85a0828c43#diff-daa5fea3872a516a483020907c6131367cbb3aacdc61e601205ac2b8a22448a3R22

File: tests/main.cpp (hunk 12 above)
Comment (Lioncat2002) [reply tests/main.cpp:64]: yum will update this test after the confirmTransaction has been implmented properly

**Pull request merged**